lib/gc/Analysis/MatmulConfigAnalysis.cpp

```diff
@@ -0,0 +1,468 @@
+                   oprandDimType[1])) {
+      if (dimType == DimType::N) {
+        N *= s;
+      } else if (dimType == DimType::K) {
+        K *= s;
+      }
+    }
+
+    // innermost Block, if the layout is blockied layout, the innermost block
+    // will derived from the layout directly
+    uint32_t defaultBlock = 32;
+    config.innerMostMBlock = M % defaultBlock == 0 ? defaultBlock : M;
+    config.innerMostNBlock = N % defaultBlock == 0 ? defaultBlock : N;
+    config.innerMostKBlock = K % defaultBlock == 0 ? defaultBlock : K;
+    SmallVector<uint32_t> givenInnermostBlock;
+    if (MDimTypeIdx.size() > 1) {
+      config.innerMostMBlock = 1;
+      for (size_t i = 1UL; i < MDimTypeIdx.size(); i++) {
+        config.innerMostMBlock *=
+            linalgOp.getShape(linalgOp.getDpsInputOperand(0))[MDimTypeIdx[i]];
+      }
+      givenInnermostBlock.push_back(config.innerMostMBlock);
+    } else {
+      givenInnermostBlock.push_back(0);
+    }
+    if (NDimTypeIdx.size() > 1) {
+      config.innerMostNBlock = 1;
+      for (size_t i = 1UL; i < NDimTypeIdx.size(); i++) {
+        config.innerMostNBlock *=
+            linalgOp.getShape(linalgOp.getDpsInputOperand(1))[NDimTypeIdx[i]];
+      }
+      givenInnermostBlock.push_back(config.innerMostNBlock);
+    } else {
+      givenInnermostBlock.push_back(0);
+    }
+    if (KDimTypeIdx.size() > 1) {
+      config.innerMostKBlock = 1;
+      for (size_t i = 1UL; i < KDimTypeIdx.size(); i++) {
+        config.innerMostKBlock *=
+            linalgOp.getShape(linalgOp.getDpsInputOperand(1))[KDimTypeIdx[i]];
+      }
+      givenInnermostBlock.push_back(config.innerMostKBlock);
+    } else {
+      givenInnermostBlock.push_back(0);
+    }
+
+    LLVM_DEBUG(llvm::dbgs()
+               << "M: " << M << ", N: " << N << ", K: " << K << "\n");
+
+    // try to read the config from the attributes
+    SmallVector<NamedAttribute> attrs(linalgOp->getAttrs());
+    bool hasPredefinedConfig = readConfigFromAttrs(config, attrs);
+
+    // if there is a given config, skip the cost model
+    if (!hasPredefinedConfig) {
+      LLVM_DEBUG(llvm::dbgs() << "No predefined config\n");
+      // TODO: Could add a weight or priority for cost model
+      SmallVector<std::tuple<CostModelFn, std::string, double>> costModelList =
+          {{workloadBalancedCost, "workloadBalancedCost", 1},
+           {vectorRegEfficiencyCost, "vectorRegEfficiencyCost ", -1},
+           {computationIntensityOnL2Cache, "computationIntensityOnL2Cache", -1},
+           {memoryConsumptionOnThreadCost, "memoryConsumptionOnThreadCost",
+            -1}};
+      SmallVector<uint32_t> shape = {M, N, K};
+      std::vector<MatmulConfig> configCandidates =
+          prepareConfigCandidates(root, sysDesc, shape, givenInnermostBlock);
+      for (auto [fn, name, threshold] : costModelList) {
+        configCandidates = filterConfigByCostModel(
+            configCandidates, linalgOp, shape, sysDesc, fn, 0.5, threshold);
+      }
```

Comment:
I have a little concern about this multiple level sorting. Although you have marked execution order as priority. However, in some cases, the result of these cost model pipeline maybe just identical with the single execution of last cost model. In another word, any other cost model before last one would make no sense in fact. IMO, the lower priority cost model should only be applied for each two candidates, whose results from higher priority cost model are equal.

Reply:
Yes, this is why I introduce the `preserveRatio` and `threshold`. This will filter out the bad values on the first several cost models. The rest of the candidates are at least not bad for the low-priority models.

Reply:
Ok, I see. But it may confuse me more about the definition of high and low priority, because the low-priority model has already filter out some candidates before high-priority model. What if the high model actually prefer those bad candidates in the eyes of lower model?

Reply:
Yes, the word `priority` is a little confusing indeed. In fact, every cost model will make their own choice and filter out the unacceptable config. The last cost model will pick the best from the rest candidates(or provide the candidates list if needed). In your case, there is a contradiction among these cost models. We could only balance them and pick a config that is not too bad for every model.

Reply:
That's fine. So the order of cost model list may also affect final config? Do we need to treat it as a tuning option in the future?

Reply:
Yes, the cost model and the execution order will be further tuned in the future